src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h

```diff
@@ -0,0 +1,269 @@
+      best.container = _traversalContainerMap[best.traversal];
+
+      if (_invalidConfigs.find(best) == _invalidConfigs.end()) {
+        // valid config found!
+        _currentConfig = best;
+        return true;
+      } else {
+        // invalid: dispose
+        _currentAcquisitions.erase(_currentAcquisitions.begin());
+      }
+    }
+
+    // No valid configuration. This should rarely happen.
+    AutoPasLog(debug, "Tuning could not generate a valid configuration.");
+  }
+
+  utils::ExceptionHandler::exception("BayesianSearch: Failed to sample an valid FeatureVector");
+  _currentConfig = FeatureVector();
+  return false;
+}
+
+void BayesianClusterSearch::sampleAcquisitions(size_t n, AcquisitionFunctionOption af) {
+  // create n lhs samples
+  auto continuousSamples = FeatureVector::lhsSampleFeatureContinuous(n, _rng, *_cellSizeFactors);
+
+  auto neighbourFun = [this](Eigen::VectorXi target) -> std::vector<Eigen::VectorXi> {
+    return FeatureVector::neighboursManhattan1(target, _dimRestrictions);
+  };
+
+  // calculate all acquisitions
+  _currentAcquisitions = _gaussianCluster.sampleAcquisition(af, neighbourFun, continuousSamples);
+
+  // sort vectors by acquisition
+  switch (af) {
+    case AcquisitionFunctionOption::mean:
+    case AcquisitionFunctionOption::lowerConfidenceBound:
+    case AcquisitionFunctionOption::upperConfidenceBound:
+      // min estimated value first
+      std::sort(_currentAcquisitions.begin(), _currentAcquisitions.end(),
+                [](const GaussianCluster::VectorAcquisition &va1, const GaussianCluster::VectorAcquisition &va2) {
+                  return va1.second < va2.second;
+                });
+      break;
+    case AcquisitionFunctionOption::variance:
+    case AcquisitionFunctionOption::expectedDecrease:
+    case AcquisitionFunctionOption::probabilityOfDecrease:
+      // max acquisition first
+      std::sort(_currentAcquisitions.begin(), _currentAcquisitions.end(),
+                [](const GaussianCluster::VectorAcquisition &va1, const GaussianCluster::VectorAcquisition &va2) {
+                  return va1.second > va2.second;
+                });
+      break;
+  }
+}
+
+bool BayesianClusterSearch::searchSpaceIsTrivial() const {
+  if (searchSpaceIsEmpty()) {
+    return false;
+  }
+
+  return _containerOptions.size() == 1 and (_cellSizeFactors->isFinite() && _cellSizeFactors->size() == 1) and
+         _traversalOptions.size() == 1 and _dataLayoutOptions.size() == 1 and _newton3Options.size() == 1;
+}
+
+bool BayesianClusterSearch::searchSpaceIsEmpty() const {
+  // if one enum is empty return true
+  return _containerOptions.empty() or (_cellSizeFactors->isFinite() && _cellSizeFactors->size() == 0) or
+         _traversalOptions.empty() or _dataLayoutOptions.empty() or _newton3Options.empty();
+}
+
+void BayesianClusterSearch::removeN3Option(Newton3Option badNewton3Option) {
+  std::remove(_newton3Options.begin(), _newton3Options.end(), badNewton3Option);
```

Comment:
```suggestion
  _newton3Options.erase(std::remove(_newton3Options.begin(), _newton3Options.end(), badNewton3Option),_newton3Options.end());
```
The call to std::remove would not remove anything, only resort the vector. See https://en.wikipedia.org/wiki/Erase%E2%80%93remove_idiom.

Which leads to the question: Why are  `_containerOptions`, `_traversalOptions`, `_dataLayoutOptions` and `_newton3Options` `std::vector` anyways? Shouldn't they be `std::set` because any entry in them should only occur once? Or do you need some special ordering?